wip: add structures to hold filtered stack graphs

nerdypepper opened the pull request:
the goal here is to get closer to simply deriving `Serialize` and `Deserialize` by introducing an intermediary data structure to contain the results of applying a `Filter` to a `StackGraph`. The flow might look like so:

- Serialization: `StackGraph::apply_filter(Filter)` -> `FilteredStackGraph`-> `serde::to_value(FilteredStackGraph)`
- Deserialization: `serde::from_value(_) -> FilteredStackGraph` -> `FilteredStackGraph::reconstruct(_)` -> `StackGraph`


This does not mean that the existing serialization code is obsolete, it will simply be repurposed to create the intermediary data structure.

hendrikvanantwerpen commented:
@nerdypepper I like the idea of having a single data type that will derive both serialization and deserialization code to guarantee consistency. So I think your approach is good. A few small remarks:

- How much can we control the resulting JSON format with this approach? Ideally it would stay the same, but even if it changes a bit, it should be nice JSON that other code could also consume easily.
- Regarding naming, it'd suggest we try to define the types as `json::StackGraph` etc. I think that is closer to what the type represents (a different representation of a stack graph), while the fact that it may or may not have been filtered is not really a part of the data structure.
- Related, I think those data types should just go in the `json.rs` module. (But perhaps the new module was meant to be temporary anyway, in which case ignore this remark.)

nerdypepper commented:
>that will derive both serialization and deserialization code to guarantee consistency

my thoughts exactly, we can also take full advantage of serde's attributes such as `skip`, `default` etc. to address your other concerns:

>How much can we control the resulting JSON format with this approach? Ideally it would stay the same, but even if it changes a bit, it should be nice JSON that other code could also consume easily.

my goal with the new implementation is to be completely compatible with the current format. if derived implementations deviate from the existing format, they will be replaced by custom serializers/deserializers where possible. if it so happens that data is not sufficient to reproduce a stack-graph on deser, i will consider changing the format a little.

>Regarding naming, it'd suggest we try to define the types as json::StackGraph etc. I think that is closer to what the type represents (a different representation of a stack graph), while the fact that it may or may not have been filtered is not really a part of the data structure.

agreed, most of the nomenclature is temporary, i'll change them up mirror the original type names.

>Related, I think those data types should just go in the json.rs module. (But perhaps the new module was meant to be temporary anyway, in which case ignore this remark.)

yes, it was a temporary module, just to start from a clean slate. ill be adding a few ser/de tests just to enforce that both methods (original and the one introduced in this PR) have identical serializations.

hendrikvanantwerpen commented:
Sounds like a good plan then! Don't hesitate to ping me when you have questions or you want me to review something.

nerdypepper commented:
@hendrikvanantwerpen managed to put in some work for this today. i'd be interested to hear your thoughts:
- i have introduced `serde` (optional) to `lsp_positions` for `Span`, `Position` and `Offset`
- the `filtered_stack_graph` module contains the conversion code to go from `crate::graph::StackGraph` to `crate::filtered_stack_graph::StackGraph`. 
- i have added a test to showcase ser/de, and to prevent any breakage going forward
- the `reconstruct` method is a stub function at the moment, which i am not sure how to fill, would love your input on that!

hendrikvanantwerpen commented on stack-graphs/src/filtered_stack_graph.rs:
```diff
@@ -0,0 +1,386 @@
+#[derive(Serialize, Deserialize, PartialEq, Eq, Debug)]
+#[serde(tag = "type")]
+pub enum Node {
+    #[serde(rename = "drop_scopes")]
```
Would it be possible to use `#[serde(rename_all = "snake_case")]` on `Node`? It would save annotations and type opportunities. [Docs](https://serde.rs/container-attrs.html#rename_all).

nerdypepper replied:
ah yes, completely forgot about that attribute.

this would serialize `JumpTo` as `jump_to`, whereas the [original implementation](https://github.com/github/stack-graphs/blob/2a75ad4d6e69441a0a9e05888947f47957decaea/stack-graphs/src/json.rs#L355) uses `jump_to_scope`. 
- shall we deviate from the original and use `jump_to` instead? or
- shall we keep the original nomenclature, and use a combination of `rename_all = "snake_case"` + `rename = "jump_to_scope"`?

hendrikvanantwerpen replied:
Just rename the enum case to JumpToScope, I'd say.

hendrikvanantwerpen commented on lsp-positions/src/lib.rs:
```diff
@@ -44,6 +44,7 @@ fn utf16_len(string: &str) -> usize {
 /// All of the position information that we have about a character in a source file
 #[repr(C)]
 #[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
```
That's very neat! Didn't know you could make `derive` conditional like that.

nerdypepper replied:
it is! another cool use-case is in combination with the builder/[typed-builder](https://docs.rs/typed-builder/latest/typed_builder/) crates, to do something like `#[cfg_attr(feature = "enable_builders", derive(TypedBuilder))]`, to conditionally allow the builder pattern.

hendrikvanantwerpen commented on stack-graphs/src/filtered_stack_graph.rs:
```diff
@@ -0,0 +1,386 @@
+}
+
+impl StackGraph {
+    pub fn reconstruct(&self) -> crate::graph::StackGraph {
+        todo!()
+    }
```
The pattern we usually use here is to define something like the following:
```rust
   fn load_into(&self, graph: &mut crate::graph::StackGraph) {
       // loop over self.files and add those to graph
       // loop over self.nodes and add those to graph
       // loop over self.edges and add those to graph
   }
```

This gives the user more control over which instance the data is loaded into, instead of always doing the allocation (and then maybe later copying).

If the elements are added in this order, everything should be available. If a file already exists in graph, we probably want to return an error, to prevent accidental merging of graphs.

hendrikvanantwerpen commented on stack-graphs/src/filtered_stack_graph.rs:
```diff
@@ -0,0 +1,386 @@
+}
+
+impl crate::graph::StackGraph {
+    pub fn apply_filter<'a>(&self, filter: &'a dyn Filter) -> StackGraph {
```
I would perhaps define this the other way around:

```rust
impl StackGraph {
   fn from_graph(graph: &StackGraph, filter: &'a dyn Filter) -> Self {
      // ...
   }
}
```

In this direction I would maybe call it `to_filtered_graph`, which aligns with common Rust method naming. We could have both directions, for more flexibility for the user, I guess.

nerdypepper replied:
yeah that makes sense. 

i would also like to stick a discussion about the public API in here, i was hoping the public API of `crate::graph::StackGraph` might look like:

- `crate::graph::StackGraph::to_json(&self) -> serde_json::Value`
- `crate::graph::StackGraph::to_json_with_filter(&self, filter: _) -> serde_json::Value`

orthogonally, we do not have to limit this to `to_json` at all, because we have the ability to serialize to any format of choice:

- `crate::graph::StackGraph::to_serializeable(&self) -> filtered_stack_graph::StackGraph` (which can then be passed to `serde_yaml::to_value` or `serde_json::to_value`), which might, at the call-site, look like this: `serde_json::to_value(&graph.to_serializable())`
- likewise with a `filter`

hendrikvanantwerpen commented on stack-graphs/src/json.rs:
```diff
@@ -230,6 +230,14 @@ impl<'a> StackGraph {
+        self.to_serializable_with_filter(&NoFilter)
+    }
+
+    pub fn to_serializable_with_filter(&self, f: &'a dyn Filter) -> crate::serde::StackGraph {
```
What do you think about calling this `to_serializable_filter`, similar to stdlib methods such as https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain_filter?

nerdypepper replied:
agreed, that is a lot better than the `with_` prefix.

hendrikvanantwerpen commented on stack-graphs/src/serde.rs:
```diff
@@ -0,0 +1,643 @@
+}
+
+impl StackGraph {
+    pub fn from_graph<'a>(graph: &crate::graph::StackGraph, filter: &'a dyn Filter) -> Self {
```
Similarly, we could call this `from_graph_filter` and have a simply `from_graph` implementation that passes the `NoFilter` argument.

hendrikvanantwerpen commented on stack-graphs/src/serde.rs:
```diff
@@ -0,0 +1,643 @@
+        for n in self.nodes.data.as_slice() {
+            match n {
+                Node::DropScopes { id, .. } => {
+                    let file = id.file().ok_or(no_file_data(id.local_id))?;
+                    let file_handle = graph.get_file(file).ok_or(not_found(file))?;
+                    let node_id = crate::graph::NodeID::new_in_file(file_handle, id.local_id);
```
Node IDs should be constructed a little differently. Some node IDs without the file part are indeed valid. A function similar to this would do the trick:

```rust
impl NodeID {
  to_node_id(&self, graph: &crate::graph::StackGraph) -> Result<crate::graph::NodeID, Error> {
    let Some(file) = self.file {
      let file = graph.get_file(file).ok_or(not_found(file))?;
      Ok(NodeID::new_in_file(file, self.local_id))
    } else if self.local_id == ROOT_NODE_ID {
      Ok(NodeID::root())
    } else if self.local_id == JUMP_TO_NODE_ID {
      Ok(NodeID::jump_to())
    } else {
      Err(Error::InvalidNodeID(self))
    }
  }
}
```

nerdypepper replied:
this reads a lot better, thanks! just to confirm, the only nodes that are permitted to not have a file (i.e., belong to all files) are the root and jump nodes, no?

hendrikvanantwerpen replied:
Correct.

hendrikvanantwerpen reviewed (commented):
I haven't look at everything in detail yet, but I left a few comments below on things I noticed.

nerdypepper commented:
>I like calling the module just serde. Does this make the json module obsolete? That would be nice!

i would like for the `serde` module to eventually replace the `json` module, however `json` currently does a lot more, with `PartialPaths` etc. would you rather have the entire `json` module translated into the `serde` module in this PR, or could i tackle that in a separate PR?

hendrikvanantwerpen commented:
> i would like for the serde module to eventually replace the json module, however json currently does a lot more, with PartialPaths etc. would you rather have the entire json module translated into the serde module in this PR, or could i tackle that in a separate PR?

@nerdypepper Good point, I didn't think about that! I think that would be my ultimate goal, but I agree with you that we should do this in a follow-up PR.

hendrikvanantwerpen commented:
@nerdypepper  Quick comment: I noticed you added something about tree-sitter to stack-graphs/Cargo.toml. We set things up so that stack-graphs is independent of tree-sitter. Anything tree-sitter specific goes into tree-sitter-stack-graphs. But I wouldn't expect anything tree-sitter specific for the serialization, tbh.

nerdypepper commented:
ah of course, i will get rid of that, accidentaly left it behind.

nerdypepper commented:
@hendrikvanantwerpen i am also considering moving the `Filter` trait and implementors such as `NoFilter` into the `serde` module, since they are not related to `json` per se. i'd be interested to hear your thoughts on this. 

this will break crates that make use of `stack_graphs::json::Filter`, so we could re-export `Filter` through the `json` module to preserve backwards compatibility.

hendrikvanantwerpen commented:
@nerdypepper Moving `Filter` makes sense. We are not too conservative yet about breaking changes, we are still v0 for a reason. But for the time being a re-export sounds good, and we can deal with this if/when we decide to remove `json` altogether.

nerdypepper commented:
@hendrikvanantwerpen i would consider this PR ready for review! i have created a new integration test module under `tests/it/serde.rs`, and moved any existing tests from `src/serde.rs` to there. i'd be happy to add more tests based on some of the graphs in `test_graphs` if necessary.

hendrikvanantwerpen reviewed (commented):
@nerdypepper This is looking great! I've left a couple of comments about code organization and some style issues, but this can be merged soon.

nerdypepper commented on lsp-positions/src/lib.rs:
```diff
@@ -52,9 +53,11 @@ pub struct Position {
     pub column: Offset,
     /// The UTF-8 byte indexes (within the file) of the start and end of the line containing the
     /// character
+    #[cfg_attr(feature = "serde", serde(skip))]
```
the reason i skipped these initially was to match the existing not deviate from the existing serialization implementation for `Position`, which seems to only use `line` and `column`. we can certainly include these.

nerdypepper replied:
sounds good.

nerdypepper commented on stack-graphs/src/json.rs:
```diff
@@ -200,6 +114,14 @@ impl<'a> StackGraph {
         JsonStackGraph(self, f)
     }
+
+    pub fn to_serializable(&self) -> crate::serde::StackGraph {
+        self.to_serializable_filter(&NoFilter)
+    }
+
+    pub fn to_serializable_filter(&self, f: &'a dyn Filter) -> crate::serde::StackGraph {
+        crate::serde::StackGraph::from_graph_filter(self, f)
+    }
```
agreed.

nerdypepper commented on stack-graphs/src/serde.rs:
```diff
@@ -0,0 +1,588 @@
+    fn load_files(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        // we check if any of the files we are about to introduce, are
+        // already present in this graph, to avoid accidental merges
+        if let Some(f) = self
+            .files
+            .data
+            .iter()
+            .find(|f| graph.get_file(f.as_str()).is_some())
+        {
+            return Err(Error::FileAlreadyPresent(f.to_owned()));
+        }
+
+        for f in self.files.data.iter() {
+            graph.add_file(f.as_str()).unwrap();
+        }
```
the reasoning for this structuring is because `graph.add_file` is a mutating operation. if we exit midway through the iteration with `graph.add_file(..).map_err(..)?`, the graph might be left in a partially merged state. what do you think?

hendrikvanantwerpen replied:
You're correct that we'd leave the stack graph in an undefined state. At the moment none of the code really makes an effort to prevent this for mutable operations. Given that, I'd say cleaner code is preferred here.

nerdypepper commented on stack-graphs/src/serde.rs:
```diff
@@ -0,0 +1,588 @@
+    }
+
+    fn load_nodes(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        for n in self.nodes.data.as_slice() {
```
no benefit really, in this case they are identical. `as_slice` (and `as_str`) are just more explicit than `&vec` and `&string` i suppose. i'll change this to `&_`

hendrikvanantwerpen commented:
@nerdypepper  At this point it's mostly cosmetic things that are left. I can also create a branch to merge your changes into, and then I can easily do the finishing touches before merging into main. 

I'm suggesting this partly because I feel that at this point addressing these little details via review/change cycles takes us both more time than doing the actual changes would. But if you prefer to finish it yourself, also okay with me!

nerdypepper commented:
@hendrikvanantwerpen lll go through the latest set of reviews, and i'd be happy to hand it over to you from there, does that sound good?

nerdypepper commented on stack-graphs/src/serde.rs:
```diff
@@ -0,0 +1,588 @@
+
+    fn load_nodes(&self, graph: &mut crate::graph::StackGraph) -> Result<(), Error> {
+        for n in self.nodes.data.as_slice() {
+            let handle = match n {
```
this is one change i have not incorporated. not that i dislike the proposal, but in the interest of moving the review forward.

nerdypepper commented:
@hendrikvanantwerpen i have looked through and implemented most changes, but feel free to take this forward as you please, thanks for the reviews and the help with making this happen!

hendrikvanantwerpen commented:
@nerdypepper Thank you for all the work, this is a great addition to the library! I'll be building on this to the same for paths soon. And we're always curious to hear/see what people are doing with stack graphs. See you around!

hendrikvanantwerpen commented:
There's a problem with the actions not being run. I'm going to create PR with these changes from within `stack-graphs` itself and see if that works. I already tried force-push but that didn't change anything.
